### ISSUE TITLE
Error in compiling syslog-ng, make[2]: *** [modules/dbparser/pdbtool/pdbtool] Error 1

### ISSUE THREAD

**Issue by ComputerFrek** [balabit/syslog-ng#2237]
# syslog-ng

## Version of syslog-ng

```
syslog-ng-3.17.2
```

## Platform

RHEL 7.5

# Issue

## Failure
gcc: warning: /usr/local/openssl/include: linker input file unused because linking not done
  CCLD     modules/dbparser/pdbtool/pdbtool
./lib/.libs/libsyslog-ng.so: undefined reference to `OPENSSL_sk_num'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_ex_data'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_CTX_set_error'
./lib/.libs/libsyslog-ng.so: undefined reference to `BIO_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `BIO_s_mem'
./lib/.libs/libsyslog-ng.so: undefined reference to `BIO_ctrl'
./lib/.libs/libsyslog-ng.so: undefined reference to `DH_check'
./lib/.libs/libsyslog-ng.so: undefined reference to `PKCS12_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_CTX_get_error'
./lib/.libs/libsyslog-ng.so: undefined reference to `ASN1_STRING_length'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_get_extension_flags'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_verify'
./lib/.libs/libsyslog-ng.so: undefined reference to `RAND_load_file'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_info_callback'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_check_private_key'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_use_certificate_chain_file'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509V3_EXT_d2i'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_get_peer_certificate'
./lib/.libs/libsyslog-ng.so: undefined reference to `BIO_new'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_get_subject_name'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_options'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `ERR_clear_error'
./lib/.libs/libsyslog-ng.so: undefined reference to `d2i_PKCS12_fp'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_add_client_CA'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_get_ext_by_NID'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_default_passwd_cb_userdata'
./lib/.libs/libsyslog-ng.so: undefined reference to `ERR_lib_error_string'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_NAME_get_text_by_NID'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_connect_state'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_new'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_get0_param'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_accept_state'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_read'
./lib/.libs/libsyslog-ng.so: undefined reference to `BN_get_rfc3526_prime_2048'
./lib/.libs/libsyslog-ng.so: undefined reference to `RAND_bytes'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_set_fd'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_session_id_context'
./lib/.libs/libsyslog-ng.so: undefined reference to `DH_new'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_use_PrivateKey'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_NAME_print_ex'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_verify_cert_error_string'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_get_ext'
./lib/.libs/libsyslog-ng.so: undefined reference to `OPENSSL_sk_value'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_load_verify_locations'
./lib/.libs/libsyslog-ng.so: undefined reference to `ERR_get_error'
./lib/.libs/libsyslog-ng.so: undefined reference to `ERR_func_error_string'
./lib/.libs/libsyslog-ng.so: undefined reference to `DH_set0_pqg'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_options'
./lib/.libs/libsyslog-ng.so: undefined reference to `ERR_reason_error_string'
./lib/.libs/libsyslog-ng.so: undefined reference to `PEM_read_bio_DHparams'
./lib/.libs/libsyslog-ng.so: undefined reference to `RAND_file_name'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_VERIFY_PARAM_set_flags'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_use_PrivateKey_file'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_default_passwd_cb'
./lib/.libs/libsyslog-ng.so: undefined reference to `RAND_write_file'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_get_ex_data'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_clear_options'
./lib/.libs/libsyslog-ng.so: undefined reference to `BN_dec2bn'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_CTX_get_current_cert'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_CTX_get_ex_data'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_set_cipher_list'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_new'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_use_certificate'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_write'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_get_cert_store'
./lib/.libs/libsyslog-ng.so: undefined reference to `TLS_client_method'
./lib/.libs/libsyslog-ng.so: undefined reference to `OPENSSL_sk_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_get_ex_data_X509_STORE_CTX_idx'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_digest'
./lib/.libs/libsyslog-ng.so: undefined reference to `BIO_new_file'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_get_issuer_name'
./lib/.libs/libsyslog-ng.so: undefined reference to `RAND_status'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_ctrl'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_CTX_get_error_depth'
./lib/.libs/libsyslog-ng.so: undefined reference to `DH_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `ASN1_STRING_get0_data'
./lib/.libs/libsyslog-ng.so: undefined reference to `BN_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `EVP_sha1'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_get_error'
./lib/.libs/libsyslog-ng.so: undefined reference to `SSL_CTX_free'
./lib/.libs/libsyslog-ng.so: undefined reference to `PKCS12_parse'
./lib/.libs/libsyslog-ng.so: undefined reference to `X509_STORE_add_cert'
./lib/.libs/libsyslog-ng.so: undefined reference to `TLS_server_method'
collect2: error: ld returned 1 exit status
make[2]: *** [modules/dbparser/pdbtool/pdbtool] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

## Steps to reproduce

1. Compiling syslog-ng

2. ...

## Compilation Env
LD_LIBRARY_PATH=/usr/local/openssl/lib; export LD_LIBRARY_PATH 
LDFLAGS="-L/usr/local/openssl/lib"; export LDFLAGS 
CPPFLAGS="-I/usr/local/openssl/include"; export CPPFLAGS 
OPENSSL_CFLAGS=/usr/local/openssl/include;export OPENSSL_CFLAGS 
OPENSSL_LIBS=/usr/local/openssl/lib;export OPENSSL_LIBS 

## Compilation Options

./configure --prefix=/usr/local/syslog-ng --enable-debug --enable-mongodb --enable-json --enable-systemd --enable-java --enable-java-modules --enable-smtp --with-libcurl=/usr/local/curl/ --with-libesmtp=/usr/local/libesmtp --with-libmongo-client=system --with-ld-library-path=/usr/local/openssl/lib

## Input and output logs (if possible)
N/A

**Comment by Kokan** [comment]
Thanks for reporting it, was it okay in the `3.16.1` release ?

**Comment by ComputerFrek** [comment]
Hi, 

Thank you for your reply.
Did not try with 3.16.1. 

I managed to compile syslog-ng with RHEL 7.5's distributed openssl. 

I was trying to compile syslog-ng with OpenSSL 1.1.0h using the following flags:
LD_LIBRARY_PATH=/usr/local/openssl/lib; export LD_LIBRARY_PATH 
LDFLAGS="-L/usr/local/openssl/lib"; export LDFLAGS 
CPPFLAGS="-I/usr/local/openssl/include"; export CPPFLAGS 
OPENSSL_CFLAGS=/usr/local/openssl/include;export OPENSSL_CFLAGS 
OPENSSL_LIBS=/usr/local/openssl/lib;export OPENSSL_LIBS 

./configure --prefix=/usr/local/syslog-ng --enable-debug --enable-mongodb --enable-json --enable-systemd --enable-java --enable-java-modules --enable-smtp --with-libcurl=/usr/local/curl/ --with-libesmtp=/usr/local/libesmtp --with-libmongo-client=system --with-ld-library-path=/usr/local/openssl/lib:/usr/local/json-c/lib

**Comment by gaborznagy** [comment]
Hi @ComputerFrek,

Do you still have the compilation problem or were you able to resolve it?
I have created and setup a clean build environment similar to yours with openssl 1.1.0h.

* I think instead of `--with-libmongo-client` you should use the `--with-mongoc` configure option (the previous one does not exist). This seems to be a doc bug.

* At first I have found that `OPENSSL_CFLAGS` misses the -I compiler flag.
After that loggen has build errors:
```
/usr/bin/ld: cannot find /usr/local/openssl/lib: File format not recognized
collect2: error: ld returned 1 exit status
Makefile:11220: recipe for target 'tests/loggen/loggen' failed
```
Are you aware that if you use `with-ld-library-path` it will result that syslog-ng will be compiled with an enabled environment wrapper and using the given value as LD_LIBRARY_PATH?
If I remove the `--with-ld-library-path` then I see the same error as you do.
```
./lib/.libs/libsyslog-ng.so: undefined reference to `TLS_server_method'
collect2: error: ld returned 1 exit status
Makefile:10897: recipe for target 'modules/dbparser/pdbtool/pdbtool' failed
make[2]: *** [modules/dbparser/pdbtool/pdbtool] Error 1
```
I need to investigate the issue more.

**Comment by gaborznagy** [comment]
And I forgot to mention that I think you should set `PKG_CONFIG_PATH`, or have you?

**Comment by ComputerFrek** [comment]
Let me try again and get back.

**Comment by gaborznagy** [comment]
I have opened a PR to remove `--with-libmongo-client` compile option.
https://github.com/balabit/syslog-ng-ose-guides/pull/37

**Comment by gaborznagy** [comment]
I'm closing this ticket due to inactivity, please send a comment if we should reopen it.
